File: 백준/Gold/9019. DSLR/DSLR.py
# 9019 DSLR
import sys
from collections import deque

def DSLR(num, command):
    if command==0:  # D
        return num*2%10000, 'D'

    if command==1:  # S
        if num==0:
            return 9999, 'S'
        return num - 1, 'S'

    if command==2:  # L
        num = num % 1000 * 10 + num//1000
        return num, 'L'

    if command==3:  # R
        temp = num % 10  # 천의자리로 올라가는것
        temp2 = num // 10 # 나머지
        num = temp*1000 + temp2
        return num, 'R'

def bfs(start_node, end_node):
    queue = deque([start_node])
    visited = [0]*10001  # 방문 집합 생성
    visited[start_node] ^= 1
    path = ['']*10001  # 최단 경로

    while queue:
        curr_node = queue.popleft()

        if curr_node == end_node:  # 탐색 종료시 최단 경로와 최단 거리 리턴
            return path[curr_node]

        for i in range(4):
            next_node, command = DSLR(curr_node, i)
            if visited[next_node] == 0:
                visited[next_node] ^= 1
                queue.append(next_node)
                # 다음 노드까지의 경로: [현재경로] + 다음 노드
                path[next_node] = path[curr_node] + command

T = int(input())
for _ in range(T):
    A, B = map(int, sys.stdin.readline().split())
    print(bfs(A, B))

File: 백준/Gold/9019. DSLR/DSLR.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

/*

*/

pair<int, char> DSLR(int num, int command) {
    if (command == 0) {
        return { (num * 2) % 10000, 'D' };
    }

    if (command == 1) {
        if (num == 0) {
            return { 9999, 'S' };
        }
        return { num - 1, 'S' };
    }

    if (command == 2) {
        num = num % 1000 * 10 + num / 1000;
        return { num, 'L' };
    }

    if (command == 3) {
        int temp = num % 10;
        int temp2 = num / 10;
        num = temp * 1000 + temp2;
        return { num, 'R' };
    }
}

string bfs(int start_node, int end_node) {
    queue<int> q;
    q.push(start_node);
    
    vector<bool> visited(10001, false);
    visited[start_node] = true;

    vector<string> path(10001, "");

    while (!q.empty()) {
        int curr_node = q.front();
        q.pop();

        if (curr_node == end_node) {
            return path[curr_node];
        }

        for (int i = 0; i < 4; ++i) {
            // auto [next_node, command] = DSLR(curr_node, i);
            auto result = DSLR(curr_node, i);
            int next_node = result.first;
            char command = result.second;
            if (visited[next_node] == false) {
                visited[next_node] = true;
                q.push(next_node);
                path[next_node] = path[curr_node] + command;
            }
        }
    }

    return ""; // If no path is found
}

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    int T;
    cin >> T;

    while(T--) {
        int A, B;
        cin >> A >> B;
        cout << bfs(A, B) << '\n';
    }

    return 0;
}
